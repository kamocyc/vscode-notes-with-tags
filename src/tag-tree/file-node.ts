class FileNode {
  // The name that should be displayed when viewing in the ui
  public displayName: string;
  // The absolute filepath
  public filePath: string;
  // The tags that are in the file itself;
  public tags: string[];
  public pathToNode: string;
  public key: string;

  constructor(
    key: string,
    filePath: string,
    pathToNode: string,
    tags: string[],
    displayName: string
  ) {
    this.displayName = displayName;
    this.key = key;
    this.filePath = filePath;
    this.pathToNode = `${pathToNode}/${this.key}`;
    this.tags = tags;
  }
}

interface IFileNodeSort {
  (fileA: FileNode, fileB: FileNode): number;
}

const fileNodeSort: IFileNodeSort = (fileA: FileNode, fileB: FileNode) => {
  if (fileA.filePath < fileB.filePath) {
    return -1;
  } else if (fileA.filePath > fileB.filePath) {
    return 1;
  } else {
    return 0;
  }
};

export { FileNode, fileNodeSort };
