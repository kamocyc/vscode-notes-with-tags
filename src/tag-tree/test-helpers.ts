import * as util from "util"; // has no default export;

// @ts-ignore
function replaceMapWithObject(key, value) {
  if (value instanceof Map || value instanceof Set) {
    var obj = {};
    for (let [mapKey, mapValue] of value) {
      // @ts-ignore
      obj[mapKey] = mapValue;
    }
    return obj;
  }
  return value;
}

// @ts-ignore
export function walkFromPath(path: string, tree) {
  const treeObject = convertTreeToObject(tree);

  const pathParts = path.split("/");
  const file = pathParts.pop();
  const tagPath = pathParts;

  // @ts-ignore
  const finalTagNode = tagPath.reduce(
    (node, nodePath) => node.tags[nodePath],
    treeObject.root
  );

  // @ts-ignore
  return finalTagNode.files[file];
}

// @ts-ignore
export function createDeepSnapshot(object) {
  return util.inspect(object, { depth: Infinity });
}

// @ts-ignore
export function convertTreeToObject(objectToSerialize) {
  return JSON.parse(JSON.stringify(objectToSerialize, replaceMapWithObject));
}
