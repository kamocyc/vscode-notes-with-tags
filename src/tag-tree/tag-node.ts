import { FileNode } from "./file-node";

class TagNode {
  public tags: Map<string, TagNode>;
  public files: Map<string, FileNode>;
  public parent: TagNode | null;

  public displayName: string;
  public tag: string;
  public pathToNode: string;

  constructor(
    parent = null,
    tag: string,
    pathToNode: string,
    displayName = ""
  ) {
    this.parent = parent;
    this.displayName = displayName;
    this.files = new Map<string, FileNode>();
    this.pathToNode = pathToNode;
    this.tag = tag;
    this.tags = new Map<string, TagNode>();
  }

  public addTag(tag: string, pathToNode: string, displayName = tag) {
    // @ts-ignore
    this.tags.set(tag, new TagNode(this, tag, pathToNode, displayName));
  }

  public getTag(tag: string) {
    return this.tags.get(tag);
  }

  public deleteTag(tag: string) {
    this.tags.delete(tag);
  }

  public hasFile(key: string) {
    return this.files.has(key);
  }

  public addFile(node: FileNode) {
    this.files.set(node.key, node);
  }

  public deleteFile(fileKey: string) {
    this.files.delete(fileKey);
  }
}

interface ITagNodeSort {
  (tagA: TagNode, tagB: TagNode): number;
}
const tagNodeSort: ITagNodeSort = (tagA: TagNode, tagB: TagNode) => {
  if (tagA.tag < tagB.tag) {
    return -1;
  } else if (tagA.tag > tagB.tag) {
    return 1;
  } else {
    return 0;
  }
};

export { TagNode, tagNodeSort };
