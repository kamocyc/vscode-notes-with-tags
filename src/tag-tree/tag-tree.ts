import { TagNode } from "./tag-node";
import { FileNode } from "./file-node";

class TagTree {
  public root: TagNode;
  private fileIndex: Map<string, TagNode[]>;

  constructor() {
    this.root = new TagNode(null, "", "");
    this.fileIndex = new Map();
  }

  /// TODO:(bdietz) - figure out if this method supports adding a file multiple times
  public addFile(filePath: string, tags: string[], displayName: string) {
    // TODO: (bdietz) - should it be the file node's job to write its own key?
    for (const tag of tags) {
      const newNode = new FileNode(
        this.createKeyForFile(filePath),
        filePath,
        tag,
        tags,
        displayName
      );
      this.addFileNode(tag, newNode);
    }
  }

  public deleteFile(filePath: string): void {
    const fileKey = this.createKeyForFile(filePath);
    if (!this.fileIndex.has(fileKey)) {
      return;
    }

    const tags = this.fileIndex.get(fileKey);

    for (const tag of tags!) {
      tag.deleteFile(fileKey);
      this.deletePathToFile(tag);
    }
    // After removing all of the nodes in the tree, remove the entry in the file index
    this.fileIndex.delete(fileKey);
  }

  public getTagsForFile(filePath: string) {
    const fileKey = this.createKeyForFile(filePath);
    if (!this.fileIndex.has(fileKey)) {
      return new Set();
    }
    const tagNodes = this.fileIndex.get(fileKey)!;
    return tagNodes.reduce((tags, tagNode) => {
      return tags.add(tagNode.tag);
    }, new Set());
  }

  public getNode(nodePath: string) {
    // @ts-ignore
    return nodePath.split("/").reduce((currentNode, pathPart) => {
      // Must be looking at a file
      if (pathPart.includes(".")) {
        return currentNode.files.get(this.createKeyForFile(pathPart));
      } else {
        return currentNode.tags.get(pathPart);
      }
    }, this.root);
  }

  private addFileNode(tagPath: string, fileNode: FileNode): void {
    /**
     * Given a tag filePath, the tags that lead up to the file may not exist in the tag tree continue down the filePath
     * building the parts of the tree that don't exist one node at a time
     */
    const nodeToAddFileTo = tagPath.split("/").reduce(
      ({ currentNode, currentPath }, pathToAdd) => {
        const newCurrentPath =
          currentPath === "" ? pathToAdd : `${currentPath}/${pathToAdd}`;

        if (currentNode.getTag(pathToAdd) !== undefined) {
          return {
            currentNode: currentNode.getTag(pathToAdd)!,
            currentPath: newCurrentPath
          };
        }
        // The node has not had this tag added yet so make sure one exists
        else {
          currentNode.addTag(pathToAdd, newCurrentPath);
          return {
            currentNode: currentNode.getTag(pathToAdd)!,
            currentPath: newCurrentPath
          };
        }
      },
      { currentNode: this.root, currentPath: "" }
    ).currentNode;

    /**
     * At this point, the tag tree has been built and we can just add the file to the
     * tag tree
     */
    nodeToAddFileTo.addFile(fileNode);
    if (!this.fileIndex.has(fileNode.key)) {
      this.fileIndex.set(fileNode.key, [nodeToAddFileTo]);
    } else {
      const nodesForFile = this.fileIndex.get(fileNode.key)!;
      this.fileIndex.set(fileNode.key, [...nodesForFile, nodeToAddFileTo]);
    }
  }

  private createKeyForFile(filePath: string): string {
    return filePath.replace(/\//g, "|");
  }

  private deletePathToFile(node: TagNode) {
    let currentNode = node;

    while (currentNode!.files.size === 0 && currentNode!.tags.size === 0) {
      const { parent } = currentNode;
      // We're not at the root of the tree yet
      if (parent !== null) {
        parent!.deleteTag(currentNode.tag);
        // @ts-ignore
        currentNode = parent;
      // At this point we've hit the root of the tree
      } else {
        break;
      }
    }
  }
}

export { TagTree };
