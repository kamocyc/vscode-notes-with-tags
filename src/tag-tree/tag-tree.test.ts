import { TagTree } from "./tag-tree";
import { createDeepSnapshot } from "./test-helpers";

describe("TagTree", () => {
  describe("deleteFile", () => {
    test("Delete node under root", () => {
      const tagTree = new TagTree();
      const filePath = "/Users/test/foo.md";
      tagTree.addFile(filePath, ["hello"], "foo.md");
      tagTree.deleteFile(filePath);

      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
    test("Delete file under shared tag", () => {
      const tagTree = new TagTree();
      const filePath = "/Users/test/foo.md";
      tagTree.addFile(filePath, ["hello"], "foo.md");
      tagTree.addFile("/Users/test/bar.md", ["hello"], "bar.md");
      tagTree.deleteFile(filePath);

      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });

    test("Delete file under shared tag with empty node", () => {
      const tagTree = new TagTree();
      const filePath = "/Users/test/foo.md";
      tagTree.addFile(filePath, ["hello/world/asuh"], "foo.md");
      tagTree.addFile("/Users/test/bar.md", ["hello"], "bar.md");
      tagTree.deleteFile(filePath);
      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });

    test("Move tag tree from one top level node to another", () => {
      const tagTree = new TagTree();
      const filePath = "/Users/test/foo.md";
      tagTree.addFile(filePath, ["hello/world"], "foo.md");
      tagTree.deleteFile(filePath);
      tagTree.addFile(filePath, ["goodbye/world"], "foo.md");
      const tag = tagTree.root.getTag("hello");
      const renamedTag = tagTree.root.getTag("goodbye");

      expect(tag).toBeUndefined();
      expect(renamedTag).toBeDefined();
    });
  });
  describe("getTagsForFile", () => {
    test("Retrieves all tags for a file", () => {
      const tagTree = new TagTree();
      tagTree.addFile("foo.md", ["hello", "world"], "foo.md");
      const tags = tagTree.getTagsForFile("foo.md");
      expect([...tags]).toEqual(["hello", "world"]);
      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
  });

  describe("addNode", () => {
    test("Single depth path", () => {
      const tagTree = new TagTree();
      tagTree.addFile("foo.md", ["hello"], "foo.md");

      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
    test("Two separate branches", () => {
      const tagTree = new TagTree();
      tagTree.addFile("foo.md", ["hello", "world"], "foo.md");

      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
    test("Two directories deep", () => {
      const tagTree = new TagTree();
      tagTree.addFile("foo.md", ["hello/world"], "foo.md");
      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
    test("Two files under the same tag", () => {
      const tagTree = new TagTree();
      tagTree.addFile("foo.md", ["hello"], "foo.md");
      tagTree.addFile("bar.md", ["hello"], "foo.md");

      expect(createDeepSnapshot(tagTree)).toMatchSnapshot();
    });
  });
});
