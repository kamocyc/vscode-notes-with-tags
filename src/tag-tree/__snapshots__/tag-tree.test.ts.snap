// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagTree addNode Single depth path 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'hello/foo.md',
               tags: [ 'hello' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} } } },
  fileIndex:
   Map {
     'foo.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags: Map { 'hello' => [Circular] } },
         displayName: 'hello',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/foo.md',
              tags: [ 'hello' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} } ] } }"
`;

exports[`TagTree addNode Two directories deep 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files: Map {},
          pathToNode: 'hello',
          tag: 'hello',
          tags:
           Map {
             'world' => TagNode {
               parent: [Circular],
               displayName: 'world',
               files:
                Map {
                  'foo.md' => FileNode {
                    displayName: 'foo.md',
                    key: 'foo.md',
                    filePath: 'foo.md',
                    pathToNode: 'hello/world/foo.md',
                    tags: [ 'hello/world' ] } },
               pathToNode: 'hello/world',
               tag: 'world',
               tags: Map {} } } } } },
  fileIndex:
   Map {
     'foo.md' => [ TagNode {
         parent:
          TagNode {
            parent:
             TagNode {
               parent: null,
               displayName: '',
               files: Map {},
               pathToNode: '',
               tag: '',
               tags: Map { 'hello' => [Circular] } },
            displayName: 'hello',
            files: Map {},
            pathToNode: 'hello',
            tag: 'hello',
            tags: Map { 'world' => [Circular] } },
         displayName: 'world',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/world/foo.md',
              tags: [ 'hello/world' ] } },
         pathToNode: 'hello/world',
         tag: 'world',
         tags: Map {} } ] } }"
`;

exports[`TagTree addNode Two files under the same tag 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'hello/foo.md',
               tags: [ 'hello' ] },
             'bar.md' => FileNode {
               displayName: 'foo.md',
               key: 'bar.md',
               filePath: 'bar.md',
               pathToNode: 'hello/bar.md',
               tags: [ 'hello' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} } } },
  fileIndex:
   Map {
     'foo.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags: Map { 'hello' => [Circular] } },
         displayName: 'hello',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/foo.md',
              tags: [ 'hello' ] },
            'bar.md' => FileNode {
              displayName: 'foo.md',
              key: 'bar.md',
              filePath: 'bar.md',
              pathToNode: 'hello/bar.md',
              tags: [ 'hello' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} } ],
     'bar.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags: Map { 'hello' => [Circular] } },
         displayName: 'hello',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/foo.md',
              tags: [ 'hello' ] },
            'bar.md' => FileNode {
              displayName: 'foo.md',
              key: 'bar.md',
              filePath: 'bar.md',
              pathToNode: 'hello/bar.md',
              tags: [ 'hello' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} } ] } }"
`;

exports[`TagTree addNode Two separate branches 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'hello/foo.md',
               tags: [ 'hello', 'world' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} },
        'world' => TagNode {
          parent: [Circular],
          displayName: 'world',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'world/foo.md',
               tags: [ 'hello', 'world' ] } },
          pathToNode: 'world',
          tag: 'world',
          tags: Map {} } } },
  fileIndex:
   Map {
     'foo.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags:
             Map {
               'hello' => [Circular],
               'world' => TagNode {
                 parent: [Circular],
                 displayName: 'world',
                 files:
                  Map {
                    'foo.md' => FileNode {
                      displayName: 'foo.md',
                      key: 'foo.md',
                      filePath: 'foo.md',
                      pathToNode: 'world/foo.md',
                      tags: [ 'hello', 'world' ] } },
                 pathToNode: 'world',
                 tag: 'world',
                 tags: Map {} } } },
         displayName: 'hello',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/foo.md',
              tags: [ 'hello', 'world' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} },
       TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags:
             Map {
               'hello' => TagNode {
                 parent: [Circular],
                 displayName: 'hello',
                 files:
                  Map {
                    'foo.md' => FileNode {
                      displayName: 'foo.md',
                      key: 'foo.md',
                      filePath: 'foo.md',
                      pathToNode: 'hello/foo.md',
                      tags: [ 'hello', 'world' ] } },
                 pathToNode: 'hello',
                 tag: 'hello',
                 tags: Map {} },
               'world' => [Circular] } },
         displayName: 'world',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'world/foo.md',
              tags: [ 'hello', 'world' ] } },
         pathToNode: 'world',
         tag: 'world',
         tags: Map {} } ] } }"
`;

exports[`TagTree deleteFile Delete file under shared tag 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             '|Users|test|bar.md' => FileNode {
               displayName: 'bar.md',
               key: '|Users|test|bar.md',
               filePath: '/Users/test/bar.md',
               pathToNode: 'hello/|Users|test|bar.md',
               tags: [ 'hello' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} } } },
  fileIndex:
   Map {
     '|Users|test|bar.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags: Map { 'hello' => [Circular] } },
         displayName: 'hello',
         files:
          Map {
            '|Users|test|bar.md' => FileNode {
              displayName: 'bar.md',
              key: '|Users|test|bar.md',
              filePath: '/Users/test/bar.md',
              pathToNode: 'hello/|Users|test|bar.md',
              tags: [ 'hello' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} } ] } }"
`;

exports[`TagTree deleteFile Delete file under shared tag with empty node 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             '|Users|test|bar.md' => FileNode {
               displayName: 'bar.md',
               key: '|Users|test|bar.md',
               filePath: '/Users/test/bar.md',
               pathToNode: 'hello/|Users|test|bar.md',
               tags: [ 'hello' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} } } },
  fileIndex:
   Map {
     '|Users|test|bar.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags: Map { 'hello' => [Circular] } },
         displayName: 'hello',
         files:
          Map {
            '|Users|test|bar.md' => FileNode {
              displayName: 'bar.md',
              key: '|Users|test|bar.md',
              filePath: '/Users/test/bar.md',
              pathToNode: 'hello/|Users|test|bar.md',
              tags: [ 'hello' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} } ] } }"
`;

exports[`TagTree deleteFile Delete node under root 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags: Map {} },
  fileIndex: Map {} }"
`;

exports[`TagTree getTagsForFile Retrieves all tags for a file 1`] = `
"TagTree {
  root:
   TagNode {
     parent: null,
     displayName: '',
     files: Map {},
     pathToNode: '',
     tag: '',
     tags:
      Map {
        'hello' => TagNode {
          parent: [Circular],
          displayName: 'hello',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'hello/foo.md',
               tags: [ 'hello', 'world' ] } },
          pathToNode: 'hello',
          tag: 'hello',
          tags: Map {} },
        'world' => TagNode {
          parent: [Circular],
          displayName: 'world',
          files:
           Map {
             'foo.md' => FileNode {
               displayName: 'foo.md',
               key: 'foo.md',
               filePath: 'foo.md',
               pathToNode: 'world/foo.md',
               tags: [ 'hello', 'world' ] } },
          pathToNode: 'world',
          tag: 'world',
          tags: Map {} } } },
  fileIndex:
   Map {
     'foo.md' => [ TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags:
             Map {
               'hello' => [Circular],
               'world' => TagNode {
                 parent: [Circular],
                 displayName: 'world',
                 files:
                  Map {
                    'foo.md' => FileNode {
                      displayName: 'foo.md',
                      key: 'foo.md',
                      filePath: 'foo.md',
                      pathToNode: 'world/foo.md',
                      tags: [ 'hello', 'world' ] } },
                 pathToNode: 'world',
                 tag: 'world',
                 tags: Map {} } } },
         displayName: 'hello',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'hello/foo.md',
              tags: [ 'hello', 'world' ] } },
         pathToNode: 'hello',
         tag: 'hello',
         tags: Map {} },
       TagNode {
         parent:
          TagNode {
            parent: null,
            displayName: '',
            files: Map {},
            pathToNode: '',
            tag: '',
            tags:
             Map {
               'hello' => TagNode {
                 parent: [Circular],
                 displayName: 'hello',
                 files:
                  Map {
                    'foo.md' => FileNode {
                      displayName: 'foo.md',
                      key: 'foo.md',
                      filePath: 'foo.md',
                      pathToNode: 'hello/foo.md',
                      tags: [ 'hello', 'world' ] } },
                 pathToNode: 'hello',
                 tag: 'hello',
                 tags: Map {} },
               'world' => [Circular] } },
         displayName: 'world',
         files:
          Map {
            'foo.md' => FileNode {
              displayName: 'foo.md',
              key: 'foo.md',
              filePath: 'foo.md',
              pathToNode: 'world/foo.md',
              tags: [ 'hello', 'world' ] } },
         pathToNode: 'world',
         tag: 'world',
         tags: Map {} } ] } }"
`;
